### ISSUE TITLE
Customer scope configuration does not function in Spack Envrionments

### ISSUE THREAD

**Issue by paulbry** [spack/spack#11008]
Though I can leveraging the custom configuration scope (`--configure-scope`) normally if I activate a Spack environment it is no longer recognized.

### Steps to reproduce the issue

In this example I used a simple `config.yaml` and have observed the same behavior with other files declared using this method.

```console
$ cat tst-cfg/config.yaml 
config:
  debug: true
  checksum: false
$ spack env activate test
$ spack -C test-config/ config blame config | grep debug
_builtin                                                      debug: False
$ despacktivate
$ spack -C test-config/ config blame config | grep debug
test-config/config.yaml:2                                     debug: True
...
```

### Error Message

There is not a definitive error message for this; however, if I run the same commands with the debug flag:

```console
$ spack -d -C tst-cfg/ config blame config
==> [2019-03-25-15:25:56.080029] Reading config file /spack/etc/spack/defaults/modules.yaml
==> [2019-03-25-15:25:56.093739] Reading config file /spack/etc/spack/defaults/linux/modules.yaml
==> [2019-03-25-15:25:56.102192] Reading config file /spack/etc/spack/defaults/config.yaml
==> [2019-03-25-15:25:56.118690] Reading config file tst-cfg/config.yaml
---                                        config:
command_line                                 debug: True
tst-cfg/config.yaml:3                        checksum: False
...
$ spack env activate test 
$ spack env status
==> In environment test
$ spack -d -C tst-cfg/ config blame config
==> [2019-03-25-16:17:24.852407] Reading config file /spack/etc/spack/defaults/modules.yaml
==> [2019-03-25-16:17:24.861573] Reading config file /spack/etc/spack/defaults/linux/modules.yaml
==> [2019-03-25-16:17:24.867582] Reading config file /spack/var/spack/environments/test/spack.yaml
==> [2019-03-25-16:17:24.872795] Reading config file /spack/etc/spack/defaults/config.yaml
---                                        config:
command_line                                 debug: True
/spack/etc/spack/defaults/config.yaml:77     checksum: True
...
```

The command line configuration scope (in this case `-d`) is in fact recognized.

### Information on your system

I have been able to reproduce this regardless of the operating system / architecture used.

### Notes

I also attempted to do some debugging regarding this issue; however, it is not definitive as I'm by no means knowledgable enough on Spack's source.

When looking at `main.py` the environment is checked before configuration scopes. This order means that `def prepare_config_scope(env):` will be executed from the `envrionment.py` file seemingly using the environments defined scope before the configuration  singleton (`def _config()`) in `config.py` is run. One potential fix seems to re-order those as such:

```python
def main(argv=None)
    ...
    if args.config_scopes:
        spack.config.command_line_scopes = args.config_scopes
    ...
    if not args.no_env:
        env = ev.find_environment(args)
```

Not sure if this would be the best solution or would present other issues.

**Comment by citibeth** [comment]
Environments come with their own config scopes.  It was not envisioned to use them with separate config scope not explicitly defined / referenced within the environment.  You can even reference an external config scope FROM an environment.

Therefore... from a design point of view, using `--configure-scope` with environments doesn't really make sense.  Are you able to use built-in environment config scopes for your purposes?

Either way, it seems the solution to this issue is to make an explicit error message explaining that `--configure-scope` is not intended to be used with environments; or else (less likely), make them work.

**Comment by paulbry** [comment]
Yeah, I make the necessary changes to the `spack.yaml` file for instance it works as intended.

**Comment by citibeth** [comment]
So.. is this issue now resolved and ready to be closed?

**Comment by paulbry** [comment]
@citibeth my issue is, so from that standpoint could be closed. Though do you think there is value in documenting it or is it already expressed in the current documentation adequately and I just missed it (completely possible)?

**Comment by paulbry** [comment]
Re-reading the documentation I can see it's not explicit in how configuration is scoped but does seem clear how I should edit configuration. I'm not sure the best approach, so if it makes sense to close this that's okay.

**Comment by citibeth** [comment]
Yes, the documentation definitely needs improving.  I couldn't figure it
out either, and I got help.  See here what I got working:

https://github.com/citibeth/spack/blob/efischer/giss2/var/spack/environments/oneway-discover/spack.yaml

**Comment by paulbry** [comment]
That works, I saw you have an issue related to this https://github.com/spack/spack/issues/10582 but I can see now I incorrectly use the `include`. I made some notes based upon your examples and think I have a better understanding of it.

Your already open issue (#10582) definitely make more sense since the only reason one might assume this to work is due to documentation not being explicit that the rules governing how Spack pulls configuration data are not observed while an environment is active.

Thanks again @citibeth for helping with this.
